cooked-validators/src/Cooked/Attack/DoubleSat.hs

```diff
@@ -0,0 +1,202 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Cooked.Attack.DoubleSat where
+
+import Control.Monad
+import Cooked.Attack.Tweak
+import Cooked.MockChain.Monad.Direct
+import Cooked.MockChain.Wallet
+import Cooked.Tx.Constraints.Type
+import Data.List
+import Data.Map (Map)
+import qualified Data.Map as Map
+import qualified Ledger as L
+import qualified Ledger.Value as L
+import Optics.Core
+import qualified PlutusTx.Numeric as Pl
+
+{- Note: What is a double satisfaction attack?
+
+A double satisfaction attack consists in trying to satisfy the requirements for
+what conceptually are two transactions in a single transaction, and doing so
+incompletely. It succeeds whenever the requirements of two validators ovelap,
+but the required outputs of the transaction are not sufficiently unique, so that
+both validators see them as satisfying "their" requirement.
+
+The mechanism is explained very well in the following analogy from the Plutus
+documentation: "Suppose that two tax auditors from two different departments
+come to visit you in turn to see if you’ve paid your taxes. You come up with a
+clever scheme to confuse them. Your tax liability to both departments is $10, so
+you make a single payment to the tax office’s bank account for $10. When the
+auditors arrive, you show them your books, containing the payment to the tax
+office. They both leave satisfied."
+
+The double satisfaction attack 'doubleSatAttack' provided by this module works
+by going through the foci of some optic on the 'TxSkel' representing the
+transaction from the left to the right, and adding some extra inputs, outputs,
+and mints depending on each focus and the current 'MockChainSt'ate. -}
+
+-- | How to combine extra constraints added for different reasons in the
+-- 'doubleSatAttack'. See the comments there.
+data DoubleSatSplitMode = TryCombinations | AllSeparate
+
+-- | A triple of transaction inputs, transaction outputs, and minted value. This
+-- is what we can add to the transaction in order to try a double satisfaction
+-- attack.
+type DoubleSatDelta = (Map SpendableOut TxSkelIn, [TxSkelOut], TxSkelMints)
+
+instance {-# OVERLAPPING #-} Semigroup DoubleSatDelta where
+  (i, o, m) <> (i', o', m') =
+    ( i <> i', -- this is left-biased union
+      o ++ o',
+      m <> m' -- see the 'Semigroup' instance of 'TxSkelMints'
+    )
```

Comment:
You might want to use [generic-deriving](https://www.stackage.org/haddock/nightly-2022-12-17/generic-deriving-1.14.2/Generics-Deriving-Semigroup.html) to avoid having to write (very boring and obvious) instances like this one. Not a requirement though, but rather JFYI!